crates/ruff/tests/lint.rs

```diff
@@ -2085,6 +2085,336 @@ select = ["UP006"]
     Ok(())
 }
 
+/// ```
+/// tmp
+/// ├── pyproject.toml #<--- no `[tool.ruff]`
+/// └── test.py
+/// ```
+#[test]
+fn requires_python_no_tool() -> Result<()> {
+    let tempdir = TempDir::new()?;
+    let project_dir = tempdir.path().canonicalize()?;
+    let ruff_toml = tempdir.path().join("pyproject.toml");
+    fs::write(
+        &ruff_toml,
+        r#"[project]
+requires-python = ">= 3.11"
+"#,
+    )?;
+
+    insta::with_settings!({
+        filters => vec![(tempdir_filter(&project_dir).as_str(), "[TMP]/"),(r"(?m)^.*(\[.*\])\[DEBUG\](.*)$","$1[DEBUG]$2")]
+    }, {
+        assert_cmd_snapshot!(Command::new(get_cargo_bin(BIN_NAME))
+            .args(STDIN_BASE_OPTIONS)
+            .arg("-v")
+            .args(["--select","UP007"])
+            .args(["--stdin-filename", "test.py"])
+            .arg("-")
+            .current_dir(project_dir)
+            .pass_stdin(r#"from typing import Union;foo: Union[int, str] = 1"#), @r###"
+        success: false
+        exit_code: 1
+        ----- stdout -----
+        test.py:1:31: UP007 [*] Use `X | Y` for type annotations
+        Found 1 error.
+        [*] 1 fixable with the `--fix` option.
+
+        ----- stderr -----
+        [ruff::resolve][DEBUG] Using Ruff default settings
+        [ruff_workspace::pyproject][DEBUG] Detected minimum supported `requires-python` version: 3.11
+        [ruff::resolve][DEBUG] Derived `target-version` from found `requires-python`
```

Comment:
It's unconventional to assert on the log messages. Could we use `--show-settings` instead? 

It may also be useful to include an assertion before that explicitly asserts that the `target-version` is not `3.11` if the `pyproject.toml` doesn't exist.

The same applies for other tests.